### PR TITLE
cifuzz: Enable SARIF reporting for systemd

### PR DESCRIPTION
**Pull request by jonathanmetzman**
Report results from CIFuzz using SARIF. 

Upload results from CIFuzz using SARIF.
This will allow CIFuzz to report issues in the security tab. 
This is a better UI than having to look through logs.
TODO(https://github.com/google/oss-fuzz/issues/10452): Add proper descriptions of UBSAN bugs.

### CODE REVIEW

**Comment by jonathanmetzman** [comment]
CC @evverx @jamacku

**Comment by jonathanmetzman** [comment]
See https://github.com/google/oss-fuzz/issues/10090

**Review by jamacku — CHANGES_REQUESTED** [review]
Please add permissions to the `Fuzzing` job:

```yml
permissions:
  security-events: write
```

They are required by `github/codeql-action/upload-sarif` GA

**Comment by yuwata** [comment]
Please extend the commit message.

**Comment by jonathanmetzman** [comment]
> Please add permissions to the `Fuzzing` job:
> 
> ```yaml
> permissions:
>   security-events: write
> ```
> 
> They are required by `github/codeql-action/upload-sarif` GA

Done.

>Please extend the commit message.

Done.

File: .github/workflows/cifuzz.yml
Comment (jamacku) [inline .github/workflows/cifuzz.yml]: Please pin to SHA `github/codeql-action/upload-sarif` GA - https://github.com/github/codeql-action/releases/tag/codeql-bundle-20230418

```suggestion
        uses: github/codeql-action/upload-sarif@9a866ed4524fc3422c3af1e446dab8efa3503411
```

**Comment by jonathanmetzman** [comment]
@jamacku done!

**Comment by jamacku** [comment]
> > Please add permissions to the `Fuzzing` job:
> > ```yaml
> > permissions:
> >   security-events: write
> > ```  
> >   
> > They are required by `github/codeql-action/upload-sarif` GA
> 
> Done.
> 

@jonathanmetzman Please create a new permission entry specific to the `Fuzzing` job. It is a practice in the systemd repo to specify minimal permissions per workflow and elevate permissions in jobs that require access to more resources. For example:

- https://github.com/systemd/systemd/blob/main/.github/workflows/differential-shellcheck.yml#L21-L22

**Comment by jonathanmetzman** [comment]
> > > Please add permissions to the `Fuzzing` job:
> > > ```yaml
> > > permissions:
> > >   security-events: write
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > They are required by `github/codeql-action/upload-sarif` GA
> > 
> > 
> > Done.
> 
> @jonathanmetzman Please create a new permission entry specific to the `Fuzzing` job. It is a practice in the systemd repo to specify minimal permissions per workflow and elevate permissions in jobs that require access to more resources. For example:
> 
> * https://github.com/systemd/systemd/blob/main/.github/workflows/differential-shellcheck.yml#L21-L22

Sorry for the misunderstanding, I think I got it now.

**Comment by jamacku** [comment]
@evverx I'm not sure if Code Scanning is/will be used in the `systemd-security` repo. FYI: It might require more permissions to use `github/codeql-action/upload-sarif` in a private repo.

Additional [permissions](https://github.com/github/codeql-action#usage):

```yml
permissions:
  # only required for workflows in private repositories
  actions: read
  contents: read
```

[GitHub Documentation on Code Scanning](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/about-code-scanning):

_Code scanning is available for all public repositories on GitHub.com. Code scanning is also available for private repositories owned by organizations that use GitHub Enterprise Cloud and have a license for GitHub Advanced Security. For more information, see "[About GitHub Advanced Security](https://docs.github.com/en/get-started/learning-about-github/about-github-advanced-security)"._

**Comment by jonathanmetzman** [comment]
I suspect this if statement prevents cifuzz from running in systemd-security: https://github.com/systemd/systemd/pull/27501/files#diff-cb0618c14c125bb79d60ba9a127ed18f4a3ce62cff9c49e26c8890afb6f35668R26

**Comment by yuwata** [comment]
CIs are unhappy:
```
Error: Path does not exist: cifuzz-sarif/results.sarif
Error: Path does not exist: cifuzz-sarif/results.sarif
    at getSarifFilePaths (/home/runner/work/_actions/github/codeql-action/9a866ed4524fc3422c3af1e446dab8efa3503411/lib/upload-lib.js:142:15)
    at Object.uploadFromActions (/home/runner/work/_actions/github/codeql-action/9a866ed4524fc3422c3af1e446dab8efa3503411/lib/upload-lib.js:137:30)
    at run (/home/runner/work/_actions/github/codeql-action/9a866ed4524fc3422c3af1e446dab8efa3503411/lib/upload-sarif-action.js:48:47)
    at async runWrapper (/home/runner/work/_actions/github/codeql-action/9a866ed4524fc3422c3af1e446dab8efa3503411/lib/upload-sarif-action.js:70:9)
```

**Comment by evverx** [comment]
Looks like the "test" directory with its weird symlinked layout confuses `shutil.copytree`:
```python
2023-05-03T14:10:20.8494231Z Traceback (most recent call last):
2023-05-03T14:10:20.8494960Z   File "/opt/oss-fuzz/infra/cifuzz/build_fuzzers_entrypoint.py", line 60, in <module>
2023-05-03T14:10:20.8546055Z     sys.exit(main())
2023-05-03T14:10:20.8546627Z   File "/opt/oss-fuzz/infra/cifuzz/build_fuzzers_entrypoint.py", line 56, in main
2023-05-03T14:10:20.8546956Z     return build_fuzzers_entrypoint()
2023-05-03T14:10:20.8547425Z   File "/opt/oss-fuzz/infra/cifuzz/build_fuzzers_entrypoint.py", line 39, in build_fuzzers_entrypoint
2023-05-03T14:10:20.8547772Z     if not build_fuzzers.build_fuzzers(config):
2023-05-03T14:10:20.8548190Z   File "/opt/oss-fuzz/infra/cifuzz/build_fuzzers.py", line 179, in build_fuzzers
2023-05-03T14:10:20.8548493Z     return builder.build()
2023-05-03T14:10:20.8548851Z   File "/opt/oss-fuzz/infra/cifuzz/build_fuzzers.py", line 142, in build
2023-05-03T14:10:20.8549125Z     if not method():
2023-05-03T14:10:20.8549526Z   File "/opt/oss-fuzz/infra/cifuzz/build_fuzzers.py", line 73, in build_image_and_checkout_src
2023-05-03T14:10:20.8549863Z     self.workspace.make_repo_for_sarif()
2023-05-03T14:10:20.8550295Z   File "/opt/oss-fuzz/infra/cifuzz/workspace_utils.py", line 98, in make_repo_for_sarif
2023-05-03T14:10:20.8550827Z     return shutil.copytree(self.repo, self.sarif)
2023-05-03T14:10:20.8551127Z   File "/usr/lib/python3.8/shutil.py", line 557, in copytree
2023-05-03T14:10:20.8551461Z     return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
2023-05-03T14:10:20.8551796Z   File "/usr/lib/python3.8/shutil.py", line 513, in _copytree
2023-05-03T14:10:20.8555518Z     raise Error(errors)
2023-05-03T14:10:20.8839572Z shutil.Error: 
```
My guess would be that passing `symlinks=True` to `shutil.copytree` should probably fix it but I have to admit I'm not sure why that directory looks like that in the first place.

**Comment by evverx** [comment]
@jonathanmetzman I'll go ahead and add the "don't merge" label. Once the "sarif-fix" branch lands I'll remove it.

**Comment by jonathanmetzman** [comment]
Landed and I've changed it back to master.

**Comment by jamacku** [comment]
It seems like there is some issue with uploading SARIF:

```
fatal: ambiguous argument 'refs/remotes/pull/27501/merge': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Could not determine current commit SHA using git. Continuing with data from user input or environment.
fatal: ambiguous argument 'refs/remotes/pull/27501/merge': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Could not determine current commit SHA using git. Continuing with data from user input or environment.
```

```js
Processing sarif files: ["cifuzz-sarif/results.sarif"]
  fatal: bad object dbfd53fa17462f556013668a920ab759c5228c48
  Failed to call git to determine merge base. Continuing with data from environment: Error: The process '/usr/bin/git' failed with exit code 128
  Error: The process '/usr/bin/git' failed with exit code 128
      at ExecState._setResult (/home/runner/work/_actions/github/codeql-action/9a866ed4524fc3422c3af1e446dab8efa3503411/node_modules/@actions/exec/lib/toolrunner.js:592:25)
      at ExecState.CheckComplete (/home/runner/work/_actions/github/codeql-action/9a866ed4524fc3422c3af1e446dab8efa3503411/node_modules/@actions/exec/lib/toolrunner.js:575:18)
      at ChildProcess.<anonymous> (/home/runner/work/_actions/github/codeql-action/9a866ed4524fc3422c3af1e446dab8efa3503411/node_modules/@actions/exec/lib/toolrunner.js:469:27)
      at ChildProcess.emit (node:events:527:28)
      at maybeClose (node:internal/child_process:1092:16)
      at Socket.<anonymous> (node:internal/child_process:451:11)
      at Socket.emit (node:events:527:28)
      at Pipe.<anonymous> (node:net:709:12)
```

**Comment by jonathanmetzman** [comment]
Ugh, let me look into this.
It's interesting that the upload succeeds though. I wonder if it just means it's unable to upload some info about the commit.
Let me see what happens if I add a crash in.

**Comment by evverx** [comment]
> It's interesting that the upload succeeds though

Looks like it's a bug in the upload-sarif action. I think it should have failed.

**Comment by evverx** [comment]
> checkout_path: cifuzz-sarif

@jamacku looks like it's related. According to https://github.com/github/codeql-action/blob/8ca5570701137b67af3d8ae3d6452f4cee6579da/upload-sarif/action.yml#L12-L15 `checkout_path` is
> The path at which the analyzed repository was checked out. Used to relativize any absolute paths in the uploaded SARIF file.

and it seems it doesn't like `cifuzz-sarif` for some reason.

Speaking of relative paths due to the way systemd is built its backtraces aren't exactly relative to the source tree in the sense that paths look like `/work/build/../../src/systemd/src/basic/process-util.c` (which among other things prevents CIFuzz from detecting affected fuzz targets). @jonathanmetzman I wonder if it can somehow affect this PR?

**Comment by jonathanmetzman** [comment]
> > checkout_path: cifuzz-sarif
> 
> @jamacku looks like it's related. According to https://github.com/github/codeql-action/blob/8ca5570701137b67af3d8ae3d6452f4cee6579da/upload-sarif/action.yml#L12-L15 `checkout_path` is
> 
> > The path at which the analyzed repository was checked out. Used to relativize any absolute paths in the uploaded SARIF file.
> 
> and it seems it doesn't like `cifuzz-sarif` for some reason.

I think I may be copying something wrong.

> 
> Speaking of relative paths due to the way systemd is built its backtraces aren't exactly relative to the source tree in the sense that paths look like `/work/build/../../src/systemd/src/basic/process-util.c` (which among other things prevents CIFuzz from detecting affected fuzz targets). @jonathanmetzman I wonder if it can somehow affect this PR?

I think it will have some negative impact, let me prevent that in my next fix.

**Comment by jonathanmetzman** [comment]
Great, I think I fixed the issue. If the sarif upload succeeds on this run I think we can merge 🤞

**Comment by jonathanmetzman** [comment]
> I think it would make sense to trigger a crash somewhere just in case to exercise that code path too.

ACK. Let's wait for this to complete and then I'll do it.

**Comment by evverx** [comment]
I forgot to mention where the fuzz targets are. Sorry! I think a NULL pointer dereference (or something like that) in https://github.com/systemd/systemd/blob/main/src/fuzz/fuzz-json.c should suffice.

**Comment by jonathanmetzman** [comment]
How does this UAF look? https://github.com/systemd/systemd/pull/27501/commits/4b4432ee555ef782fe424043f3b99eea7ec11fc9

**Comment by evverx** [comment]
If it fails fast It can be caught by the bad build check though so maybe it should be skipped for the time being. Let's see.

**Comment by jonathanmetzman** [comment]
I think this will fail fast but not too fast. Hopefully you don't have too many fuzzers that it has enough time.

**Comment by jonathanmetzman** [comment]
I need to fix another thing. This time I have a pretty clear idea what to do right off thebat.

**Comment by evverx** [comment]
@jonathanmetzman as far as I understand CIFuzz doesn't emit locations (and it could be that it has something to do with unusual systemd backtraces but I'm not sure). Since the bot keeps removing the labels on force-pushes I'll go ahead and convert this PR to draft to make it clear that it's a work in progress.

**Comment by jonathanmetzman** [comment]
> @jonathanmetzman as far as I understand CIFuzz doesn't emit locations (and it could be that it has something to do with unusual systemd backtraces but I'm not sure). Since the bot keeps removing the labels on force-pushes I'll go ahead and convert this PR to draft to make it clear that it's a work in progress.

Yeah I'm having some trouble with this. I'm gonna do some unittests to fix this problem instead of my current strategy.

**Comment by evverx** [comment]
If that issue is caused by systemd backtraces I'm not sure it makes much sense to fix it given that it's the only project where backtraces look like that. `systemd` should fix its stuff instead. (it would be useful regardless of SARIF because it would make it possible for CIFuzz to detect fuzz targets affected by PRs).

**Comment by jonathanmetzman** [comment]
> If that issue is caused by systemd backtraces I'm not sure it makes much sense to fix it given that it's the only project where backtraces look like that. `systemd` should fix its stuff instead. (it would be useful regardless of SARIF because it would make it possible for CIFuzz to detect fuzz targets affected by PRs).

I think it's different than this actually, though that may also be an issue :-\

**Comment by jonathanmetzman** [comment]
So I think the issues I've been having here are the result of (lib)ClusterFuzz substituting the filename instead of LLVMFuzzerTestOneInput when it's in the stack. I think I should handle that case gracefully now. I'll also check that I can handle a real crash after this.

**Comment by jonathanmetzman** [comment]
Nice it works!

**Comment by jonathanmetzman** [comment]
The last steps are to merge my change on the oss-fuzz side and then change this action to use oss-fuzz's master.

**Comment by evverx** [comment]
Looking at the alert I wonder if it's possible to embed backtraces in alerts?

**Comment by jonathanmetzman** [comment]
> Looking at the alert I wonder if it's possible to embed backtraces in alerts?

I think we can put a few frames there. I'm going to do that soonish.

This can land after https://github.com/google/oss-fuzz/pull/10349 lands

**Comment by evverx** [comment]
I think it would make sense to inject as many frames as possible to make it easier to figure out what fails. This way it wouldn't be necessary to view the action logs.

Also even though "Don't crash" is motivational :-) my understanding (based on a quick look at the part responsible for SARIF) is that the action should translate backtraces into bug classes. Could it be that it couldn't recongize that crash somehow?

**Comment by jonathanmetzman** [comment]
> I think it would make sense to inject as many frames as possible to make it easier to figure out what fails. This way it wouldn't be necessary to view the action logs.

I'm planning on doing this.

> 
> Also even though "Don't crash" is motivational :-) my understanding (based on a quick look at the part responsible for SARIF) is that the action should translate backtraces into bug classes. Could it be that it couldn't recongize that crash somehow?

Yeah it must not recognize the crash since this is my fallback. I'll look into that.

**Comment by evverx** [comment]
I'm puzzled as to why it produced only that one alert. The ASan and UBSan jobs should have uploaded two different SARIF files pointing to the same location as far as I understand. I wonder if GitHub somehow merged them or the last file won? I have to admit I'm not sure how several jobs uploading SARIF in parallel work.

**Comment by jonathanmetzman** [comment]
I think this is ready to land.
The reason "Don't crash" was reported is because it is the fallback. I haven't catagorized the UBSAN bugs yet but I plan od doing this as well as adding stackframes (If possible, which I think it is).

File: .github/workflows/cifuzz.yml
Comment (evverx) [inline .github/workflows/cifuzz.yml]: @jonathanmetzman could you pin it to 0225834cc549ee0ca93cb085b92954821a145866 to make it consistent with the other workflows where `upload-sarif` is used? (Dependabot tends to be confused by the same actions in different workflows pointing to different commits). It got bumped to that commit in https://github.com/systemd/systemd/pull/27879/files#diff-649a84adf982ddf68273f1bb0de882a20e97b06632b5432a6a1504a3f23cb2e6 as far as I can see.

Could you also squash the commits?

Other than that the PR looks good to me. If it looks good to @jamacku as well I think it should be good to go.

**Comment by evverx** [comment]
> I haven't catagorized the UBSAN bugs yet but I plan od doing this as well as adding stackframes 

Could you open an OSS-Fuzz/CIFuzz issue and refer to it in the commit message? It should hopefully make it easier to keep track of those features.

File: .github/workflows/cifuzz.yml
Comment (evverx) [reply .github/workflows/cifuzz.yml]: > Other than that the PR looks good to me

On second thought I think it would maybe be better to bring backtraces first. Without them alerts are kind of cryptic. Dunno. I'm on the fence.

**Review by jamacku — APPROVED** [review]
LGTM

Thank you, @jonathanmetzman.

**Comment by jonathanmetzman** [comment]
@evverx Done

**Comment by evverx** [comment]
Thanks!

Looks like https://github.com/systemd/systemd/pull/27501#discussion_r1213531078 hasn't been addressed yet but I have to admit I'm not sure whether Dependabot is still confused by different hashes (sometimes GitHub fixes its bugs). I'll go ahead and add the "good-to-merge-with-minor-suggestions" label and let @jamacku decide and press the "merge" button :-)

**Comment by jonathanmetzman** [comment]
> Thanks!
> 
> Looks like [#27501 (comment)](https://github.com/systemd/systemd/pull/27501#discussion_r1213531078) hasn't been addressed yet but I have to admit I'm not sure whether Dependabot is still confused by different hashes (sometimes GitHub fixes its bugs). I'll go ahead and add the "good-to-merge-with-minor-suggestions" label and let @jamacku decide and press the "merge" button :-)

Sorry, I just addressed it.

**Comment by jamacku** [comment]
Thank you, I'm going to merge it. CI failure seems unrelated to this PR.

**Comment by jonathanmetzman** [comment]
Thank you!

**Pull request merged**
